### ISSUE TITLE
Support regex for ServiceRole spec.rules.paths

### ISSUE THREAD

**Issue by wereeder** [istio/istio#16585]
(This is used to request new product features, please visit <https://discuss.istio.io> for questions on using Istio)

**Describe the feature request**
Support regex paths for ServiceRole spec.rules.paths, similar to how the Policy supports regex for spec.peers.jwt.trigger_rules.excluded_paths

**Describe alternatives you've considered**
Paths currently support wildcard patching for suffix, prefix or exact match, but not a mix of these. I tried using wildcard at beginning and end but this did not work. Without regex support, I have had to define several variations for allowing access to the swagger-ui.html. For example, I have added the following:

```yaml
apiVersion: rbac.istio.io/v1alpha1
kind: ServiceRole
metadata:
  name: view-swagger
  namespace: dp
spec:
  rules:
  - services: ["*"]
    paths:
    - "*/swagger-ui.html"
    - "*/api-docs"
    - "*/swagger-resources"
    - "*/swagger-resources/configuration/ui"
    - "*/swagger-resources/configuration/security"
    - "*/springfox-swagger-ui/fonts/open-sans-v15-latin-regular.woff2"
    - "*/springfox-swagger-ui/fonts/open-sans-v15-latin-700.woff2"
    - "*/springfox-swagger-ui/fonts/source-code-pro-v7-latin-300.woff2"
    - "*/springfox-swagger-ui/fonts/titillium-web-v6-latin-700.woff2"
    - "*/springfox-swagger-ui/springfox.js?v=2.9.2"
    - "*/springfox-swagger-ui/swagger-ui-bundle.js?v=2.9.2"
    - "*/springfox-swagger-ui/swagger-ui-standalone-preset.js?v=2.9.2"
    - "*/springfox-swagger-ui/springfox.css?v=2.9.2"
    - "*/springfox-swagger-ui/swagger-ui.css?v=2.9.2"
    - "*/springfox-swagger-ui/favicon-32x32.png?v=2.9.2"
    - "*/springfox-swagger-ui/favicon-16x16.png?v=2.9.2"
```

I would like to simplify this with a regex, such as:
```yaml
apiVersion: rbac.istio.io/v1alpha1
kind: ServiceRole
metadata:
  name: view-swagger
  namespace: dp
spec:
  rules:
  - services: ["*"]
    regex_paths:
    - \/.*swagger.*
    - \/.*api-docs
```

**Affected product area (please put an X in all that apply)**

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[X] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Additional context**

**Comment by jpatters** [comment]
Instead of introducing a new key (`regex_paths`) I think it would be nice to add some consistency to the apis and use `StringMatch` like in `VirtualService` and `Gateway`.
ie:
```
paths:
- exact: /v1
- regex: "\/.*/docs\/.*"
- prefix: /v2
```

**Comment by edgel** [comment]
This feature will be very useful to define the authz policy, without this feature the only way that I can image is to use the Mixer Policy with OPA (it is powerful but seems not a recommend way), so I suggest AuthorizationPolicy to support another header matcher mode (istio/pilot/pkg/security/authz/model/matcher/header.go):
- /exact: for exact string mach
- */suffix: for suffix string match
- prefix/*: for prefix string match
- *: for non-empty string match
- **|^/regex/.*$|**: **for regex string match**

```
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: httpbin-1
  namespace: foo

spec:
  selector:
    matchLabels:
      app: httpbin
      version: v1
  rules:
    - from:
        - source:
            principals: ["principal", "principal-prefix-*", "*-suffix-principal", "*"]
            requestPrincipals: ["requestPrincipals", "requestPrincipals-prefix-*", "*-suffix-requestPrincipals", "*"]
            namespaces: ["ns", "ns-prefix-*", "*-ns-suffix", "*"]
            ipBlocks: ["1.2.3.4", "5.6.0.0/16"]
      to:
        - operation:
            methods: ["method", "method-prefix-*", "*-suffix-method", "*"]
            hosts: ["exact.com", "*.suffix.com", "prefix.*", "*"]
            ports: ["80", "90"]
            paths: ["/exact", "/prefix/*", "*/suffix", "*", "|^/regex/.*$|"]
```

**Comment by GODBS** [comment]
This becomes important in Istio 1.5 now that the alpha Authentication Policy is being replaced with the Request Authentication and Peer Authentication. There is no other way to exclude paths then to use an Authorization Policy which does not allow regex.

**Comment by incfly** [comment]
Hmm indeed this seems a feature regression... I confirmed we don't have regex in the beta API (both req/authz). @yangminzhu you must know more on this. We do not include regex in the authz api is that for a specific reason or due to our negligence?

**Comment by yangminzhu** [comment]
yes, we dropped the regex support in the beta RequestAuthentication and didn't add it in the AuthorizationPolicy as we're not sure how many users will actually need the regex and what's the API should be.

> Instead of introducing a new key (regex_paths) I think it would be nice to add some consistency to the apis and use StringMatch like in VirtualService and Gateway.
ie:

paths:
- exact: /v1
- regex: "\/.*/docs\/.*"
- prefix: /v2

@jpatters thanks for the suggestion, we ever considered this a little bit but @liminw has some concerns that this will make the policy quite redudant.

> /exact: for exact string mach
*/suffix: for suffix string match
prefix/*: for prefix string match
*: for non-empty string match
|^/regex/.*$|: for regex string match

@edgel this seems a more easier and incremental way to support the regex, but we need to think about the detail syntax. 

@diemtvu @liminw It looks like many users are actually wanting the regex match in authZ policy, could you share some thoughts here, maybe we could support it like this: `paths: ["/exact", "/prefix*", "*-suffix", "*", "regex(...)"]`?

**Comment by myidpt** [comment]
We usually don't make a feature request as a P0. Previsouly this was an experimental feature and we don't expect users to use it in production.

**Comment by nak3** [comment]
+1
Knative on Istio have to use `triggerRules.excludedPaths.{exact,prefix}` for authentication policy as knative service gets health checking and metrics collection from external(knative system pods). So this feature is really necessary.
Cloud Run's solution also suggested it [here](https://cloud.google.com/solutions/authorizing-access-to-cloud-run-on-gke-services-using-istio#configuring_istio_authentication) for now.

**Comment by incfly** [comment]
>  use triggerRules.excludedPaths.{exact,prefix} for authentication policy

new authorization policy has `not_paths` since 1.5.

**Comment by lgarrett-isp** [comment]
> We usually don't make a feature request as a P0. Previsouly this was an experimental feature and we don't expect users to use it in production.

@myidpt 

What is the recommended solution for the following REST pattern of sub-resources?

Suppose I build a CRUD REST service for _Countries_. Each country has a sub-resource _City_. Following REST convention, I believe this means paths like
/countries
/countries/{country id}
/countries/{country id}/cities
/countries/{country id}/cities/{city id}

I have a rule to ALLOW requests to `/countries/MyCountry*` if request.auth.claims["countries"] contains "MyCountry".

Some users who are authorized for `/countries/{any country}` are NOT authorized to POST cities (they are read-only users). My natural thought is to create a rule to DENY `POST` for `/countries/*/cities/*` when request.auth.claims["readonly"] is true.

But without regex support I cannot see how to enable what I assumed was a very common situation for REST services. 

I am not an expert REST developer but I had thought this was an extremely common pattern--can you help me understand why this pattern is uncommon (what are the better patterns for subresources?), or what led to the conclusion that users would not use this regex support for this purpose in production?

**Comment by DanSalt** [comment]
> I am not an expert REST developer but I had thought this was an extremely common pattern--can you help me understand why this pattern is uncommon (what are the better patterns for subresources?), or what led to the conclusion that users would not use this regex support for this purpose in production?

@lgarrett-isp 

Can't speak for everyone, but it's definitely the same issue for us. We've followed what we believe is the correct REST convention, and find ourselves unable to express this within the policy. We're trying to replace our custom Envoy-based sidecar with a full Istio implementation, but the lack of this support means we'll likely end up with some kind of hybrid, which is unfortunate. We *definitely* need this for Production use-cases.

**Comment by lgarrett-isp** [comment]
Yep, we have run head-first into the same; we are moving away from Istio for our REST services until it has more support for basic REST conventions.

**Comment by howardjohn** [comment]
@yangminzhu is this solved by https://github.com/istio/istio/issues/27984?

**Comment by yangminzhu** [comment]
> @yangminzhu is this solved by #27984?

yes, but we're evaluating if glob matching would be good enough for most use cases.

---

@wereeder  @jpatters  @edgel  @nak3  @lgarrett-isp  @DanSalt @ajit-hybris  @aguromano @ajitchahal  @mcieplak @rolandkool 

I'm wondering would the [glob matching](https://en.wikipedia.org/wiki/Glob_(programming)) good enough for your use cases? It looks like most examples (e.g. `\/.*swagger.*` or `/countries/*/cities/*`) posted here can be supported by glob matching.  I feel there should be some use cases that requires the full regex but not sure detail examples, do you have any examples for this? Or you think the glob matching is indeed good enough? Thanks

**Comment by anannaya** [comment]
@yangminzhu  For My header is be like x-auth-request-groups : 'CN=EMSUDEMY,OU=Groups,O=cco.xxx.com, CN=all-xxxx-people,OU=xxxx Groups,O=cco.xxxx.com, CN=xxxxxxxxx,OU=xxxx Groups,O=cco.xxxx.com'

now I would like match *all-xxxx-people*.
Glob matching should be fine for my usecase.

**Comment by FENGXIANGLI** [comment]
there seems a discussion about this in the community concerning about the security risk @yangminzhu do you have any updates for this regex supporting?

**Comment by williamaronli** [comment]
there seems a discussion about this in the community concerning about the security risk @yangminzhu do you have any updates for this regex supporting?

**Comment by mstrYoda** [comment]
Is there any update for this issue? Will you support glob matching?

**Comment by previousdeveloper** [comment]
+1

**Comment by carnei-ro** [comment]
Hey guys, I'm running Istio 1.9.0.
I tried to implement the countries/cities example:

> Allow POST to /v1/countries/{countries_id}
> Deny POST to /v1/countries/{countries_id}/cities/{city_id}

```yaml
  action: ALLOW
  rules:
      to:
        - operation:
            methods:
              - POST
            paths:
              - /v1/countries/*
            notPaths:
              - /v1/countries/*/cities/*
```

I can not make the example work as expected. I could find this:
`curl -X POST http://iam.consumer--account/v1/countries/baz/cities/foo`: the request hits my service (**NOT** the behavior I was expecting)
`curl -X POST http://iam.consumer--account/v1/countries/*/cities/foo`: the request is blocked by Envoy. It seems the "wildcard" is only accepted at the end. 

IMHO The issue should remain open... It is not possible to implement "REST" policy in this way :(

**Comment by mstrYoda** [comment]
> Hey guys, I'm running Istio 1.9.0.
> I tried to implement the countries/cities example:
> 
> > Allow POST to /v1/countries/{countries_id}
> > Deny POST to /v1/countries/{countries_id}/cities/{city_id}
> 
> ```yaml
>   action: ALLOW
>   rules:
>       to:
>         - operation:
>             methods:
>               - POST
>             paths:
>               - /v1/countries/*
>             notPaths:
>               - /v1/countries/*/cities/*
> ```
> 
> I can not make the example work as expected. I could find this:
> `curl -X POST http://iam.consumer--account/v1/countries/baz/cities/foo`: the request hits my service (**NOT** the behavior I was expecting)
> `curl -X POST http://iam.consumer--account/v1/countries/*/cities/foo`: the request is blocked by Envoy. It seems the "wildcard" is only accepted at the end.
> 
> IMHO The issue should remain open... It is not possible to implement "REST" policy in this way :(
> 
> Is it possible to allow writing a condition like `safe_regex` of envoy?

Actually it is possible to write regex based path matching in envoy. We can write regex path matching as below in envoy:
![image](https://user-images.githubusercontent.com/12763626/109866254-1bd71e00-7c76-11eb-8c47-acd83699db9c.png)

Under the hood Istio uses almost the same configuration, all wildcard matchings uses regex under the hood: https://github.com/istio/istio/blob/master/pilot/pkg/security/authz/matcher/string.go#L29

It is very simple to support regex in AuthorizationPolicy. We can specify path matching regex as a field of custom resource or we can give path with prefix like "regex: /some/path/*/".

I can create a pr for that if Istio maintainers consider to support it.

**Comment by shaneqld** [comment]
I've run into this problem while migrating from Istio on GKE (1.4.10) to 1.9.1.  In my case, the application required JWT auth on `/v1beta1/forms/*` except for a few "custom methods", e.g. `/v1beta1/forms/*:submit`.

In this case, `/v1beta1/forms/*:submit` is a subset of `/v1beta1/forms/*`, so glob matching isn't ideal.  Personally I like to be explicit and I think regex is needed for that.

I like the proposal raised by @edgel as it doesn't require changing the API like using StringMatch would.

With regards to the format of the string, some thoughts:

- I assume something as suggested like `|regex|` or `regex: regex` would be very unlikely to break existing rules.
- Because wildcards can currently only be used at the start or end of strings, `*` could possibly be used to signify a matcher mode, give room for extensibility and not cause backwards compatibility problems. For example: `r*` prefix for regex (`r*^/my/.*/regex$`) or `g*` prefix for glob (`g*/my/*/glob`).

**Comment by ognjen-it** [comment]
+1 bump

It would be good if we implement regex!

**Comment by LEDfan** [comment]
Could this be reopened? I still see the value of this feature and would like to use it. Thanks!

**Comment by yangminzhu** [comment]
reopened the issue for more discussions and also unassigned myself as I'm not actively working on this.

**Comment by mstrYoda** [comment]
@yangminzhu Hi, I am volunteer to implement this feature, we really need it on production, can you assign it to me? Also we can work with volunteers who wants to be pair.

**Comment by yangminzhu** [comment]
@mstrYoda sorry for the late reply, just assigned the issue to you, thanks for helping on this!

**Comment by johnbuluba** [comment]
Hello, 
We are currently blocked by this issue as we have the need to match header values using regex.

@mstrYoda Is there any progress on this? Is there any way we can help solving this issue?

Also there seems to be another issue regarding the same subject [#25021], which one is the main issue that tracks the progress? Maybe we can close one of them as duplicate (cc @yangminzhu).

Thanks

**Comment by mstrYoda** [comment]
@johnbuluba Thank you for reaching me, I will create a pr for this soon, we heavily need this in production too :)

**Comment by yangminzhu** [comment]
@mstrYoda thanks for the help.  Just want to let you know we tried to solve this before but there are some difficulties that we do not have good solutions, you might need to think about these (or write a short proposal) before sending the actual PR:

1. it's hard to support this without changing the API, technically we could introduce special syntax to represent the regex in the existing field but we want to avoid that as it is not backward compatible and generally a bad user experience in reusing the same field for different purposes.

2. some general concern about the regex's security and usability, most users probably only need wildcard matching.

**Comment by mstrYoda** [comment]
> @mstrYoda thanks for the help. Just want to let you know we tried to solve this before but there are some difficulties that we do not have good solutions, you might need to think about these (or write a short proposal) before sending the actual PR:
> 
> 1. it's hard to support this without changing the API, technically we could introduce special syntax to represent the regex in the existing field but we want to avoid that as it is not backward compatible and generally a bad user experience in reusing the same field for different purposes.
> 2. some general concern about the regex's security and usability, most users probably only need wildcard matching.

Hi @yangminzhu actually I just thinking to support it with same field by "regex:" prefix but like I said, it might bad user experience. I agree with u, probably supporting only wildcard matching is what actually needed here (to support child resource paths for restful apis).

**Comment by magoon** [comment]
Please add wildcard matching within the string, to augment the existing support for prefix/suffix matching. We just need to be able to handle routes like `/order/*/details`. I know there are a lot of requests for regular expressions, which would be nice, but please don't let that stop us from having basic wildcard support to Authorization Policy Rules.

**Comment by mstrYoda** [comment]
@johnbuluba @magoon @ognjen-it Hi guys I've made the development for wildcard matching. I am going to build Istio locally and test the functionality and create the pr 🙏

**Comment by mikebryant** [comment]
To share our use-case (on AuthorizationPolicy conditions)

We're using [oauth2-proxy](https://github.com/oauth2-proxy/oauth2-proxy) with [external authentication](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/), which provides a comma-separated list of groups in the `X-Auth-Request-Groups` header

With only prefix & suffix matching, we can't specify a condition of "I need a user to have this group", and we need this.

Regex matching (as provided for in istio/api#2173) would solve this for us :)

**Comment by mikebryant** [comment]
I'd be happy to avoid regex, if istio could support a way of matching elements of comma-separated headers ([RFC2616](https://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html#sec4.2))

Though I would note (for our case at least), this is a header that is guaranteed to be present by the authenticating proxy, so it's from a trusted source. (And the input data is from our own SSO system, which is also trusted). So this isn't a value controllable by an attacker.

**Comment by nTahiri** [comment]
Any input this issue. Is there a way to handle regex for path variable manipulation.
Please advise

**Comment by Harmelodic** [comment]
Also have a use case for wanting regex/wildcards inside path definitions:

```
/path/to/something/<id>
/path/to/something/<id>/sub/thing
/path/to/something/<id>/sensitive/thing
```

Which can be mitigated right now by using `notPaths` with a suffix, but it's quite messy and unintuitive, and runs the risk of blocking access to paths that accidentally match with the `notPaths` suffix rule.

As for implementation desire: Just full-blown regex would be nice as it'd mean it's not another pseudo-regex specification that I need to learn as a developer.

**Comment by hihellobolke** [comment]
Is the regex matching support via https://github.com/istio/api/pull/2173 still under consideration? This would have really helped us too!

**Comment by mstrYoda** [comment]
Anyone can make comment on my last question on that review?

**Comment by dacloutier** [comment]
Can we please have support for path parameter rewrites in istio? The fix to support this has been available for so long and all that's needed is a merge.

**Comment by frankjkelly** [comment]
:-(

**Comment by dacloutier** [comment]
It's been 3 years with no progress... and the last comment from istio was: "Regex matching a complex seems ripe for a bypass."

:,(

**Comment by mstrYoda** [comment]
It looks like lots of users need that functionality (also we need that in production too), so why not move forward on my pr: https://github.com/istio/istio/pull/35641/ ?

Is there any Istio Security team member here? If anyone knows, can mention them here?

**Comment by sschepens** [comment]
@mstrYoda we also need this, and I can contribute, I had another [PR](https://github.com/istio/api/pull/2173) for solving this, though that is only API change, i'm still pending istiod changes.

I can help you making a document to express the use case and changes necessary, also attendingo to a security meeting to help push this.

**Comment by mstrYoda** [comment]
> @mstrYoda we also need this, and I can contribute, I had another [PR](https://github.com/istio/api/pull/2173) for solving this, though that is only API change, i'm still pending istiod changes.
> 
> 
> 
> I can help you making a document to express the use case and changes necessary, also attendingo to a security meeting to help push this.

Of course that will be nice to collaborate 🙏

**Comment by howardjohn** [comment]
I would encourage us to look into whether _regex_ specifically is needed, or blog/pattern matching. Regex has a huge surface area. If all of the use cases are for APIs like `/customer/{id}/blah`, then something matching that form may make more sense

**Comment by dacloutier** [comment]
> I would encourage us to look into whether _regex_ specifically is needed, or blog/pattern matching. Regex has a huge surface area. If all of the use cases are for APIs like `/customer/{id}/blah`, then something matching that form may make more sense

The use case that is most needed is as described above... If something other than "regex" for matching makes more sense, than I'm sure that would make everyone here very happy.

**Comment by teddy-wahle** [comment]
I also need support for `/customer/{id}/blah` and may be unable to use AuthorizationPolicy if this is not supported.

**Comment by sschepens** [comment]
I think most use cases should be covered by patterns `/customer/{id}/blah` but since we allow regex in virtual service routing, it makes sense to provide the same type of matching in AuthorizationPolicy.

**Comment by emwalker** [comment]
I'm guessing something less heavy than regexes will do the trick for most people here.  The `/customer/{id}/blah` suggestion would be adequate in my own case.  This issue seems like a big one that is worth prioritizing, as routes with ids in the middle are pretty common.  I'm now looking at a case where something like `POST /tenants/{id}/application_forms/guest` needs to be allowed through without authorization.  Not sure what I'm going to do here.

**Comment by emwalker** [comment]
I think a suitable workaround in my case is to modify the external authorization service to return "authorized" for routes matching the path with the embedded id.

**Comment by jgournet** [comment]
Last comments from this ticket and closed https://github.com/istio/istio/issues/25021 are about having a strong community demand
=> has there been any update on this ?

**Comment by howardjohn** [comment]
This is not implemented due to lack of demand, but due to a tentative decision to avoid regex in authz where possible due to security, complexity, and performance concerns. @aryan16 has been writing up a design for an alternative using globbing (eg `/customer/{id}/blah`) and should attach the design doc here once its ready.

**Comment by nberkley-gp** [comment]
Note that this doesn't just come up with path processing, but also when matching on things like the XFCC header and want to check for some element out of the middle (say, a SAN) while not having to change your auth policy every time the cert rotates. We don't need full regex for that either, but we at least need unanchored matches (i.e. `*foo*`) and not a solution that only works for paths.

**Comment by hexiaodai** [comment]
+1

**Comment by coarsehorse** [comment]
+1

**Comment by radubutnaru** [comment]
> This is not implemented due to lack of demand, but due to a tentative decision to avoid regex in authz where possible due to security, complexity, and performance concerns. @aryan16 has been writing up a design for an alternative using globbing (eg `/customer/{id}/blah`) and should attach the design doc here once its ready.

We are facing the same issue, glob style matching would do the trick. Would love to see this implemented. Was there any progress on the globbing-based design?

**Comment by naimadswdn** [comment]
@aryan16 is there any update regarding design documentation for the glob style matching in the AuthorizationPolicy? 
Seems like that is a highly demanding feature.

**Comment by nberkley-gp** [comment]
@howardjohn and/or @aryan16 any word on that design that was mentioned back in March? The lack of unanchored matches essentially makes XFCC support useless in Authorization Policies. Would a PR adding support for matches "contains" matches (i.e. `*FOO*`) be acceptable in the mean time if I were to write one?

**Comment by BobakMariusz** [comment]
+1

**Comment by alekberovvusal** [comment]
+1

**Comment by mojeico** [comment]
+1

**Comment by michalgut** [comment]
+1

**Comment by mmf55** [comment]
+1

**Comment by sajalg09** [comment]
+1

**Comment by kddsultan** [comment]
+1

**Comment by loeffel-io** [comment]
+1

**Comment by rokkarinn** [comment]
+1

**Comment by Lanchez** [comment]
+1

**Comment by mattjnewberry** [comment]
+1. We'd really appreciate this feature

**Comment by caibirdme** [comment]
Here's the user case:
we have such URLs:
- /user/{:id}/hobby
- /user/{:id}/profile
- ...
we want most of the requests should be Authzed except for  /user/{:id}/hobby. Without regexp, we cannot achieve this goal! The only way is to implement this logic in authz pod, but I don't like that anyway. because if you do that, all the functions in AuthzPolicy.rule can be done in authz pod as well.

**Comment by DanielRosenkranz** [comment]
+1

**Comment by smaspe** [comment]
Without some form of dynamic matching this feature is usable in only the simplest of cases (for all the reasons given above)

**Comment by pintonunes** [comment]
+1

**Comment by KSmigielski** [comment]
@aryan16 @howardjohn Hey. We are migrating from our in-house service-mesh to Istio. And I have some data for you. We have around 1,800 services of which almost 1,200 use the equivalent of AuthorizationPolicy.

From 1,200 services over 300 use regexps. Our service-mesh only supports regexps for path matching, and at this moment there is over 1,600 such rules. 

I analyzed all usage of regexps, but of course this are only estimations, because it would take a lot of work to fully analyze all policies. 

200 can be migrated to exact match. I don't know why people use regexp match at all :upside_down_face:.
100 can be easily migrated to prefix match. 
Another 400 maybe can be migrated to prefix match. This group mostly uses constructs like `/some/path/[^/]+` or `/some/path/[^/]+/[^/]`. But to be sure, I would have to perform full analyze of all policies for each service separately.
Next 900 I will call "restful API". They contain some kind of identifier inside of path, so it looks like `/some/path/[^/]+/rest/of/the/path`. Sometimes it has more than one identifier in path `/some/[^/]+/path/[^/]/again`.
Lastly there are a few which can be migrated to suffixMatch.

Additionally, in all the above groups some users restrict characters to some subset e.g. UUID characters or digits. But for me, it is over-engineering, and it could simplify to `[^/]+`.

Last but not least, when you think that you know how to use regexps and even your teammate can't spot a mistake
```
/some/path*
/some/path/{orderId}/again
/some/path/[^.].+
```

In summary, I think for our use case it would be enough to support glob matching. The best implementation for us would be, if `/some/*/path` would match to `/some/identifier/path`, but not to `/some/super/long/path`, so it would behave similar to current `[^/]+`. It would easily cover "restful API" cases and rest which can not migrate to prefix match. 

I hope this helps you make your decision. If you give the green light, we can implement it. Because it's a waste of our and your time if it never gets merged

**Comment by howardjohn** [comment]
I am 100% on board with a glob based implementation. The user facing implementation would be either `/path/{something}/bar` (where 'something' has no meaning other than documentation to the user) or `/path/*/bar`. In either case, these would not match `/path/a/b/bar` (only a single fragment would be matched).

The gap is a safe way to do this. I have opened https://github.com/envoyproxy/envoy/issues/30724 to this end. This _might_ not require envoy changes if the regex approach is deemed secure.

If we can resolve the Envoy issue I think this is very likely to move forward. I cannot speak for the whole community but I believe ^ will address any open concerns based on passed conversations

**Comment by kyessenov** [comment]
Can you use an extension to categorize your API calls and emit API attribute that is then consumed by RBAC? I think that was the intention behind https://github.com/istio-ecosystem/wasm-extensions/tree/master/extensions/attributegen since multiples regexes during matching is generally problematic.

**Comment by howardjohn** [comment]
> Can you use an extension to categorize your API calls and emit API attribute that is then consumed by RBAC? I think that was the intention behind https://github.com/istio-ecosystem/wasm-extensions/tree/master/extensions/attributegen since multiples regexes during matching is generally problematic.

attributegen is relying on regex which, IMO, has the same problems that need to be addressed by https://github.com/envoyproxy/envoy/issues/30724.

Either way, I think its important we solve this as a first class API without hoops. This is the top 2 demanded open issue (by thumbs up) and has been a constant request.

**Comment by kyessenov** [comment]
There's no need to use regex, you can port https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig to wasm, or just expose a "path classifier" filter from that. The important thing is that URL wildcards are not standardized, so we should not be dictating it in our APIs. It requires some extensibility.

**Comment by howardjohn** [comment]
Can we just expose https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig as a matcher in RBAC? That would, IMO, perfectly solve the issue

**Comment by michaelbannister** [comment]
I think that option looks great; in my org have multiple teams using Cloud Endpoints for auth who are reluctant to move to Istio AuthorizationPolicy because they have rules using those "REST-style" paths that can't be expressed with the current prefix/suffix matching.

**Comment by eduardosanzb** [comment]
Hello!
Is there something we can do to move this forward? Kind regards

**Comment by mstrYoda** [comment]
> Can we just expose https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig as a matcher in RBAC? That would, IMO, perfectly solve the issue

Hi @howardjohn it looks like the exact solution. How would you think to proceed, as a bacward compatible with the existing CRDs? Or creating a new version that takes type in new field?

I would like to work on this one as we still need it a lot in production 🙏

**Comment by howardjohn** [comment]
> > Can we just expose https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig as a matcher in RBAC? That would, IMO, perfectly solve the issue
> 
> Hi @howardjohn it looks like the exact solution. How would you think to proceed, as a bacward compatible with the existing CRDs? Or creating a new version that takes type in new field?
> 
> I would like to work on this one as we still need it a lot in production 🙏

I would imagine the same field just allows some new syntax. not sure if that is `a/*/b` or `a/{segment}/b`, etc though

**Comment by mstrYoda** [comment]
> > > Can we just expose https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig as a matcher in RBAC? That would, IMO, perfectly solve the issue
> 
> > 
> 
> > Hi @howardjohn it looks like the exact solution. How would you think to proceed, as a bacward compatible with the existing CRDs? Or creating a new version that takes type in new field?
> 
> > 
> 
> > I would like to work on this one as we still need it a lot in production 🙏
> 
> 
> 
> I would imagine the same field just allows some new syntax. not sure if that is `a/*/b` or `a/{segment}/b`, etc though

So wdyt about if I modify this pr to use UriTemplateMatchConfig (instead of regex matcher) according to Envoy doc that you sent?

Pr: https://github.com/istio/istio/pull/35641

**Comment by jasperanders** [comment]
Not sure if this fits here, but regex Matching for Headers is also something I would appreciate. If more souphisticated rule matchig is allowed in AuthorizationPolicies this might help. Maybe the Api of the VirtualService could be reused? ([Exact, Prefix, Regex](https://istio.io/latest/docs/reference/config/networking/virtual-service/#StringMatch)).

**Comment by sahityacodes** [comment]
is there support for "#" in the paths? I have made a couple of tests, but it does not seem to work. Any help?

**Comment by howardjohn** [comment]
> is there support for "#" in the paths? I have made a couple of tests, but it does not seem to work. Any help?

1. Please do not comment irrelevant comments on busy issues like this; use https://github.com/istio/istio/discussions
2. `#` should not be a part of any requests that get to Envoy, per HTTP spec. These are a client side feature of URL only. So authz policy against `#` does not make sense. https://datatracker.ietf.org/doc/html/rfc3986#section-3.5

**Comment by thiagogcm** [comment]
this is supported upstream now:
https://github.com/envoyproxy/envoy/pull/31447

**Comment by keithmattix** [comment]
Seems like we can expect it to be released in Envoy [1.30](https://www.envoyproxy.io/docs/envoy/latest/version_history/v1.30/v1.30.0). Any idea what the target release date is for that? I'm not sure istio/api#2173 is the right API though given that upstream Envoy just added a new matcher

**Comment by dmitriy-kharchenko** [comment]
According to schedule planned release date of Envoy 1.30 is 16/04/2024 - [Envoy Major release schedule](https://github.com/envoyproxy/envoy/blob/main/RELEASES.md#major-release-schedule)

**Comment by keithmattix** [comment]
Well that link's going in my bookmarks :) Thanks! I think that means that Istio 1.22 can very easily include this feature. Either me or someone on my team can tackle this if @mstrYoda is ok with it? I see they're currently assigned to the issue

**Comment by mstrYoda** [comment]
> Well that link's going in my bookmarks :) Thanks! I think that means that Istio 1.22 can very easily include this feature. Either me or someone on my team can tackle this if @mstrYoda is ok with it? I see they're currently assigned to the issue

It is okay to me, also I would like to contribute too if there is a chance for it 👍

**Comment by jaellio** [comment]
Hi @mstrYoda! I am excited to takle this work together! Here are my thoughts for a solution and how we can pair. 

Let's leave the `paths` and `notPaths` fields in the AuthorizationPolicy resource as is for backwards compatibility. We can create new fields called `pathTemplates` and `notPathTemplates` that can be an array of paths based on [uri template matching](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig) rules supported by envoy. The user can set one of `paths` or `pathTemplates` but not both. If the user wishes to set `notPathTemplates` they must not set `paths` or `pathTemplates` and vise versa.  The `pathTemplates` fields could look something like this:

```yaml
apiVersion: security.istio.io/v1beta1
kind: AuthorizationPolicy
metadata:
  name: httpbin
  namespace: foo
spec:
  action: ALLOW
  rules:
  - from:
    - source:
        principals: ["cluster.local/ns/default/sa/sleep"]
    - source:
        namespaces: ["test"]
    to:
    - operation:
        methods: ["GET"]
        pathTemplates: ["/info/*/{file}", "/info/v2/*/{file}]
    - operation:
        methods: ["POST"]
        pathTemplates: ["/data/**"]
    when:
    - key: request.auth.claims[iss]
      values: ["https://accounts.google.com"]
```

`/info/**` would match `/info/foo/bar.txt` or `info/v2/foo/bar.txt`
`/data/{file}` would match `/data/foo/bar/test`

Would you be interested in taking on the api changes? I can work on the control plane changes. Let me know your thoughts!

**Comment by michaelbannister** [comment]
> `/data/{file} would match /data/foo/bar/test`

Just checking this is a typo? :) `{name}` matches "one path segment up to the next path separator" according to the uri template doc.

I suppose it'd be more effort to only partially support the uri template syntax (i.e. not the named variables) than just passing it straight through to the Envoy config, but as a user I'm wondering if there'd be any value in the named variables _beyond_ it being a kind of documentation: might the captured values get exposed somewhere for (say) use in access logs, or an EnvoyFilter, in a `when` condition? (If not now, perhaps in future.) And would there be any performance penalty for that over the non-capturing matchers?

**Comment by mstrYoda** [comment]
> Hi @mstrYoda! I am excited to takle this work together! Here are my thoughts for a solution and how we can pair.
> 
> Let's leave the `paths` and `notPaths` fields in the AuthorizationPolicy resource as is for backwards compatibility. We can create new fields called `pathTemplates` and `notPathTemplates` that can be an array of paths based on [uri template matching](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig) rules supported by envoy. The user can set one of `paths` or `pathTemplates` but not both. If the user wishes to set `notPathTemplates` they must not set `paths` or `pathTemplates` and vise versa. The `pathTemplates` fields could look something like this:
> 
> ```yaml
> apiVersion: security.istio.io/v1beta1
> kind: AuthorizationPolicy
> metadata:
>   name: httpbin
>   namespace: foo
> spec:
>   action: ALLOW
>   rules:
>   - from:
>     - source:
>         principals: ["cluster.local/ns/default/sa/sleep"]
>     - source:
>         namespaces: ["test"]
>     to:
>     - operation:
>         methods: ["GET"]
>         pathTemplates: ["/info/*/{file}", "/info/v2/*/{file}]
>     - operation:
>         methods: ["POST"]
>         pathTemplates: ["/data/**"]
>     when:
>     - key: request.auth.claims[iss]
>       values: ["https://accounts.google.com"]
> ```
> 
> `/info/**` would match `/info/foo/bar.txt` or `info/v2/foo/bar.txt` `/data/{file}` would match `/data/foo/bar/test`
> 
> Would you be interested in taking on the api changes? I can work on the control plane changes. Let me know your thoughts!

Hi @jaellio 👋  Thanks for the collaboration. I have the same questions with the @michaelbannister 🤔 

I did not work on API side before but I have some idea and [experience on the control plane side](
https://github.com/istio/istio/pull/35641). Can we pair/collaborate on both? I can make the changes on the API side and would love to see developments on the control plane too.

**Comment by howardjohn** [comment]
I am a bit worried about having multiple ways to represent path, especially when its also in 'when'. It can be pretty complex to have 4 (with 'not' variants) rules for the same thing.

I also don't think we need to support the full syntax of Envoy's URI template. The URI template is the implementation detail of our API; we should define the API we want and map it onto the URI template. Specifically, I don't see a strong need to support `{name}` and `*`. We should support only one. `{}` are not valid in URLs, while `*` is, so in theory that may be more backwards compatible. That also is more future proof if we ever want to make assertions based on the extracted values (I don't think this is supported in Envoy, but its why its in the Envoy API I assume).

**Comment by jaellio** [comment]
> > `/data/{file} would match /data/foo/bar/test`
> 
> Just checking this is a typo? :) `{name}` matches "one path segment up to the next path separator" according to the uri template doc.
> 
> I suppose it'd be more effort to only partially support the uri template syntax (i.e. not the named variables) than just passing it straight through to the Envoy config, but as a user I'm wondering if there'd be any value in the named variables _beyond_ it being a kind of documentation: might the captured values get exposed somewhere for (say) use in access logs, or an EnvoyFilter, in a `when` condition? (If not now, perhaps in future.) And would there be any performance penalty for that over the non-capturing matchers?

@michaelbannister Yes, apologies for the typo! I forgot to update the comment :) It should be `/data/{file} would match /data/test`

I think for now it would make sense to not support named variables. I had originally assumed that you could reference the variable later on in the template or as an environment variable. Primarily, the variable seems to be a source of documentation (and confusion). We could revisit adding support for named variables later on based on user feedback and increased functionality support (either by Istio or envoy). For now, we can interpret any paths containing `{` as invalid.

**Comment by jaellio** [comment]
> I am a bit worried about having multiple ways to represent path, especially when its also in 'when'. It can be pretty complex to have 4 (with 'not' variants) rules for the same thing.

Are `paths` and `notPaths` supported [values](https://istio.io/latest/docs/reference/config/security/conditions/) for the the `when` conditional? 

If we opted to support URI template matching for the existing `paths` and `notPaths` fields we would need to interpret `*` as a templating type and as a valid character in the path. Need to think more on this, but I could see this causing some issues.

**Comment by howardjohn** [comment]
> @howardjohn Are paths and notPaths supported [values](https://istio.io/latest/docs/reference/config/security/conditions/) for the when conditional?

Ah I guess not, my bad.

I agree `*` _could_ be a valid value today so that is tricky (although, its a very weird URL). `{}` are explicitly not valid URL charecters so no one can legitimately have those which may be a good motivation to use them; it also is consistent if we later decide to have `{name}`

**Comment by jaellio** [comment]
If we only support `{}` for templating we can use the existing `paths` and `notPaths` fields. 

If we don't want to incorporate named variables (regardless of whether or not we do anything with them) we could support the following templating and map them to envoy's supported uri templating:

- {} or {*} : Matching one path segment up to the next path separator: /.
- {**} : Matching zero or more path segments. **NOTE: It isn't clear from the envoy uri templating documentation if `{**}` must be the last operator**

Alternatively, if we are interested in incorporating a named variable (but not doing anything with the variable at this time) we could support the same templating with `{` as described in the uri templating documentation.

**Comment by howardjohn** [comment]
> We will also support the existing `*` functionality:
> 
> * `*` : Matching zero or many path segments.

Isn't this "Match any character"? Its not "path aware", just equivalent to `.*` regex. While a `{}` is more like `\/.+?\/` (probably that is not a robust regex, just an example that is close).

> I find the named variables introduce unnecessary confusion and would opt for the former option. We can add support for named variables later if necessary.

+1. I would, similarly, not support `{*}` or `{**}`. Just `{}`.

**Comment by keithmattix** [comment]
> Isn't this "Match any character"? Its not "path aware", just equivalent to .* regex. While a {} is more like \/.+?\/ (probably that is not a robust regex, just an example that is close).

I think in the [UriTemplate](https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/path/match/uri_template/v3/uri_template_match.proto#extensions-path-match-uri-template-v3-uritemplatematchconfig) context, `*` becomes path aware (which is one reason why I'm more in favor of @jaellio's original [not]pathTemplate approach). 

> +1. I would, similarly, not support {*} or {**}. Just {}.

I think the OP's original example would require something like `{*}/springfox-swagger-ui/{**}`

**Comment by jaellio** [comment]
I would agree `*` is not path aware currently for Istio.
Depending on how you are defining `{}` I don't think it supports a complex enough set of use cases.

**Comment by jaellio** [comment]
I think adding new fields (`pathTemplates` and `notPathTemplates`) to support more complex templating is simpler if we want to address the asks of the OP: there are less backwards combability concerns, it's more extensible, it'd easier for users to distinguish between new vs old functionality, and it doesn't feel like we are boxing ourselves into limited templating capabilities. I think our templating support should be scoped, but by choice, not by design or backwards compatibility limitations.

**Comment by jaellio** [comment]
If we aren't in favor of adding new fields, I think we should add support for the following. I don't think `{}` alone covers enough use cases. 
- {} or {*} : Matching one path segment up to the next path separator: /.
- {**} : Matching zero or more path segments. This must be the last operator.

If the path contains `{` or `}` then we will interpret translate the path as a uri template rather than as a string match.

**Comment by michaelbannister** [comment]
This sounds reasonable to me. It'd be good if the corresponding Istio docs were clearer than the UriTemplate docs, though:
> `*` : Matches a single path component, up to the next path separator: `/`

but then this example:
> `/videos/*/*/*.m4s` would match `videos/123414/hls/1080p5000_00001.m4s`

which suggests that the `.` of the file extension can be a boundary of a `*` as well – but I suspect `/videos/*/*/*` would _also_ match that same path so maybe it's flexible?

Would we want to explicitly disallow any use of `{` or `}` which _doesn't_ match one of those specific patterns? e.g. `/orders/{id}` should fail validation

**Comment by jaellio** [comment]
> This sounds reasonable to me. It'd be good if the corresponding Istio docs were clearer than the UriTemplate docs, though:
> 
> > `*` : Matches a single path component, up to the next path separator: `/`
> 
> but then this example:
> 
> > `/videos/*/*/*.m4s` would match `videos/123414/hls/1080p5000_00001.m4s`
> 
> which suggests that the `.` of the file extension can be a boundary of a `*` as well – but I suspect `/videos/*/*/*` would _also_ match that same path so maybe it's flexible?

Totally agree. This could be a lot more clear - I'll can verify this by updating the unit tests I think. I already make some updates to verify if `{**}` had to be the last operator. https://github.com/envoyproxy/envoy/pull/33104

> 
> Another thought: Would we want to explicitly disallow any use of `{` or `}` which _doesn't_ match one of those specific patterns? e.g. `/orders/{id}` should fail validation

I think that explicitly disallowing alternative use cases might be better than an undefined outcome even if we document it as unsupported functionality.  I am not sure how complicated the conditional/regex would be, but I think it makes sense to only allow `{*}` and `{**}`. If we chose to add support for named variables in the future, we could remove the restriction on our side. 

Regarding the impact of explicitly disallowing the use of `{` if it doesn't match a supported pattern: Previously, if a user provided a path with `{` or any other combination of brackets the path would technically be invalid but wouldn't result in an error. The path would instead result in no matches since `{` is not a valid character in a path. By explicitly disallowing unsupported patterns containing `{` the user will now see an error. I am not sure if this would result in the AuthorizationPolicy not being applied, but this could unintentionally cause authorization policies to not be applied if they contain unsupported paths. Not sure if this is a realistic scenario or how many users this might impact, just wanted to call it out.

**Comment by costinm** [comment]
Added a comment on the PR - not against adding matches, but only if we understand:
- if HttpRoute can be extended to do the same
- what are the security implications - if routes and authorizations are not aligned ( the API is supposed to work as a 'vendor extension' when using HttpRoute / Gateway )
- backward compatibility risk - now that everything is v1, I don't think we can play games or start moving to v2.

**Comment by jaellio** [comment]
Closing in favor of creating a new issue to track broader adoption and integration of URI templating.
